File: python/src/contriwork_notifications/adapters/discord.py
"""Discord webhook adapter — HTTPS POST with a JSON body."""

from __future__ import annotations

from typing import Any

import httpx

from ..adapter import AdapterDeliverResult, AdapterStatus
from ..errors import ErrorCode
from ..payload import Payload
from ..severity import Severity

_DEFAULT_TIMEOUT_S = 10.0
# Discord message flag bit (since 2023): clients render the message but
# suppress push and badge notifications for it.
_SUPPRESS_NOTIFICATIONS_FLAG = 4096

_GENERIC_FAILURE = "Notification delivery failed"
_GENERIC_AUTH = "Authentication rejected"


class DiscordWebhookAdapter:
    """Discord webhook delivery via HTTPS POST.

    The webhook URL is opaque to the package (consumer's responsibility to
    load from its own config source). The body is sent as plain Markdown
    content (``**title**`` + body, optional ``[label](url)`` link); the
    package does not produce embed payloads in v0.1.0.

    Silent mode: sets the ``SUPPRESS_NOTIFICATIONS`` flag (4096) so the
    Discord client renders the message in-channel without pushing a system
    notification.
    """

    def __init__(
        self,
        webhook_url: str,
        *,
        name: str = "discord-webhook",
        timeout_seconds: float = _DEFAULT_TIMEOUT_S,
    ) -> None:
        self._webhook_url = webhook_url
        self.name = name
        self._timeout = timeout_seconds

    async def is_available(self) -> bool:
        return bool(self._webhook_url)

    async def deliver(
        self,
        severity: Severity,
        payload: Payload,
        silent: bool,
    ) -> AdapterDeliverResult:
        del severity
        content = f"**{payload.title}**\n{payload.body}"
        if payload.url is not None:
            label = payload.url_title or payload.url
            content += f"\n[{label}]({payload.url})"
        body: dict[str, Any] = {"content": content}
        if silent:
            body["flags"] = _SUPPRESS_NOTIFICATIONS_FLAG

        try:
            async with httpx.AsyncClient(timeout=self._timeout) as client:
                response = await client.post(self._webhook_url, json=body)
        except httpx.TimeoutException:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.TIMEOUT,
                detail=_GENERIC_FAILURE,
            )
        except httpx.HTTPError:
            return AdapterDeliverResult(
                status=AdapterStatus.RETRIABLE_FAILURE,
                error_code=ErrorCode.UPSTREAM_ERROR,
                detail=_GENERIC_FAILURE,
            )

        return _map_response(response.status_code)


def _map_response(status_code: int) -> AdapterDeliverResult:
    # Discord webhook returns 204 No Content on success.
    if 200 <= status_code < 300:
        return AdapterDeliverResult(status=AdapterStatus.DELIVERED)
    if status_code == 429:
        return AdapterDeliverResult(
            status=AdapterStatus.RETRIABLE_FAILURE,
            error_code=ErrorCode.RATE_LIMITED,
            detail=_GENERIC_FAILURE,
        )
    if status_code in (401, 403):
        return AdapterDeliverResult(
            status=AdapterStatus.PERMANENT_FAILURE,
            error_code=ErrorCode.AUTH_FAILED,
            detail=_GENERIC_AUTH,
        )
    if 400 <= status_code < 500:
        return AdapterDeliverResult(
            status=AdapterStatus.PERMANENT_FAILURE,
            error_code=ErrorCode.ADAPTER_INVALID_PAYLOAD,
            detail=_GENERIC_FAILURE,
        )
    return AdapterDeliverResult(
        status=AdapterStatus.RETRIABLE_FAILURE,
        error_code=ErrorCode.UPSTREAM_ERROR,
        detail=_GENERIC_FAILURE,
    )

File: typescript/src/adapters/discord.ts
import type { Adapter, AdapterDeliverResult } from "../adapter";
import { AdapterStatus } from "../adapter";
import { ErrorCode } from "../errors";
import type { Payload } from "../payload";
import type { Severity } from "../severity";

const DEFAULT_TIMEOUT_MS = 10_000;
const SUPPRESS_NOTIFICATIONS_FLAG = 4096;
const GENERIC_FAILURE = "Notification delivery failed";
const GENERIC_AUTH = "Authentication rejected";

export interface DiscordWebhookOptions {
  readonly webhookUrl: string;
  readonly name?: string;
  readonly timeoutMs?: number;
  readonly fetchImpl?: typeof fetch;
}

/**
 * Discord webhook delivery. Plain Markdown content (`**title**` + body,
 * optional `[label](url)` link). Silent mode sets the message `flags`
 * field to 4096 (`SUPPRESS_NOTIFICATIONS`).
 */
export class DiscordWebhookAdapter implements Adapter {
  readonly name: string;
  private readonly webhookUrl: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(options: DiscordWebhookOptions) {
    this.webhookUrl = options.webhookUrl;
    this.name = options.name ?? "discord-webhook";
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.fetchImpl = options.fetchImpl ?? globalThis.fetch.bind(globalThis);
  }

  async isAvailable(): Promise<boolean> {
    return Boolean(this.webhookUrl);
  }

  async deliver(
    _severity: Severity,
    payload: Payload,
    silent: boolean,
  ): Promise<AdapterDeliverResult> {
    let content = `**${payload.title}**\n${payload.body}`;
    if (payload.url !== undefined) {
      const label = payload.urlTitle ?? payload.url;
      content += `\n[${label}](${payload.url})`;
    }
    const bodyObj: Record<string, unknown> = { content };
    if (silent) {
      bodyObj["flags"] = SUPPRESS_NOTIFICATIONS_FLAG;
    }
    const body = JSON.stringify(bodyObj);

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let response: Response;
    try {
      response = await this.fetchImpl(this.webhookUrl, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body,
        signal: controller.signal,
      });
    } catch {
      if (controller.signal.aborted) {
        return {
          status: AdapterStatus.RetriableFailure,
          errorCode: ErrorCode.Timeout,
          detail: GENERIC_FAILURE,
        };
      }
      return {
        status: AdapterStatus.RetriableFailure,
        errorCode: ErrorCode.UpstreamError,
        detail: GENERIC_FAILURE,
      };
    } finally {
      clearTimeout(timer);
    }

    return mapResponse(response.status);
  }
}

function mapResponse(statusCode: number): AdapterDeliverResult {
  // Discord webhook returns 204 No Content on success; treat any 2xx.
  if (statusCode >= 200 && statusCode < 300) {
    return { status: AdapterStatus.Delivered };
  }
  if (statusCode === 429) {
    return {
      status: AdapterStatus.RetriableFailure,
      errorCode: ErrorCode.RateLimited,
      detail: GENERIC_FAILURE,
    };
  }
  if (statusCode === 401 || statusCode === 403) {
    return {
      status: AdapterStatus.PermanentFailure,
      errorCode: ErrorCode.AuthFailed,
      detail: GENERIC_AUTH,
    };
  }
  if (statusCode >= 400 && statusCode < 500) {
    return {
      status: AdapterStatus.PermanentFailure,
      errorCode: ErrorCode.AdapterInvalidPayload,
      detail: GENERIC_FAILURE,
    };
  }
  return {
    status: AdapterStatus.RetriableFailure,
    errorCode: ErrorCode.UpstreamError,
    detail: GENERIC_FAILURE,
  };
}
